MembraneRTC/src/main/java/org/membraneframework/rtc/models/TrackContext.kt

```diff
@@ -1,6 +1,26 @@
 package org.membraneframework.rtc.models
 
+import org.membraneframework.rtc.TrackEncoding
 import org.membraneframework.rtc.media.RemoteTrack
 import org.membraneframework.rtc.utils.Metadata
 
-data class TrackContext(val track: RemoteTrack?, val peer: Peer, val trackId: String, val metadata: Metadata)
+data class TrackContext internal constructor(
+    val track: RemoteTrack?,
+    val peer: Peer,
+    val trackId: String,
+    val metadata: Metadata,
+    val vadStatus: VadStatus,
+    val encoding: TrackEncoding?,
+    val encodingReason: EncodingReason?,
+    private val trackContextInternal: TrackContextInternal
```

Comment:
Now I see what was the idea of being able to set listeners on any copy of TrackContext. But is it really how people do this? How this will affect future plans on moving to events?

Reply:
> is it really how people do this?

I think the problem with this class is that it was a `data class` - that is the class designed to just store data/couple fields together. Data classes are nice for this and we treat them as immutable and I was giving it to the user and user could do anything with them and everything was nice.
But now we want to add some logic to the data class so maybe it would be better to turn it into a normal, mutable class. Kotlin is not typescript and copying object doesn't happen so easily. And restrict the user so that they can't set anything (`internal set` or sth). But that will be a bit of a breaking change. Is it ok to change?

And now I'm thinking maybe I should've add an interface for a listener and use interface instead of just lambda. Kotlin is nice and for the user it will look like a lambda but if user wishes to implement callback on one object they can do so. Something like here: https://developer.android.com/develop/ui/views/components/button#ClickListener

> How this will affect future plans on moving to events?
I guess it won't - you can add functions for adding/removing event listeners just like functions for setting a listener? What I implemented is already some kind of event listener, I just don't allow setting many event listeners.